### PR TITLE
[SYCL] Adding caching when using interop constructor

### PR DESCRIPTION
**Pull request by maximdimakov**
When interop constructor is used it creates new impl  every time. This patch allows to use already created impls in the constructor.

### CODE REVIEW

File: sycl/include/CL/sycl/platform.hpp [hunk 1]
```diff
@@ -132,6 +137,11 @@ class __SYCL_EXPORT platform {
   shared_ptr_class<detail::platform_impl> impl;
   platform(shared_ptr_class<detail::platform_impl> impl) : impl(impl) {}
 
+  static std::unordered_map<cl_platform_id,
+                            std::weak_ptr<detail::platform_impl>>
+      platform_impls;
+  static std::mutex platform_mutex;
```
Comment (alexbatashev) [inline sycl/include/CL/sycl/platform.hpp:143]: This isn't going to work properly. Both unordered_map and mutex are non-trivial classes. See [this](https://github.com/intel/llvm/blob/sycl/sycl/doc/GlobalObjectsInRuntime.md) document for more info.

File: sycl/include/CL/sycl/platform.hpp [hunk 2]
```diff
@@ -112,8 +113,12 @@ class __SYCL_EXPORT platform {
...
   template <backend BackendName>
   auto get_native() const -> typename interop<BackendName, platform>::type {
-    return reinterpret_cast<typename interop<BackendName, platform>::type>(
-        getNative());
+    auto cl_platform =
+        reinterpret_cast<typename interop<BackendName, platform>::type>(
+            getNative());
+    std::lock_guard<std::mutex> lock(platform_mutex);
+    platform_impls[cl_platform] = impl;
+    return cl_platform;
```
Comment (alexbatashev) [inline sycl/include/CL/sycl/platform.hpp:121]: Nit
```suggestion
    auto ClPlatform =
        reinterpret_cast<typename interop<BackendName, platform>::type>(
            getNative());
    std::lock_guard<std::mutex> Lock(platform_mutex);
    platform_impls[ClPlatform] = impl;
    return ClPlatform;
```

File: sycl/include/CL/sycl/device.hpp [hunk 3]
```diff
@@ -192,6 +195,10 @@ class __SYCL_EXPORT device {
   shared_ptr_class<detail::device_impl> impl;
   device(shared_ptr_class<detail::device_impl> impl) : impl(impl) {}
 
+  static std::unordered_map<cl_device_id, std::weak_ptr<detail::device_impl>>
+      device_impls;
+  static std::mutex device_mutex;
```
Comment (alexbatashev) [inline sycl/include/CL/sycl/device.hpp:200]: See my comment for the same piece in `platform` class.

File: sycl/source/device.cpp [hunk 4]
```diff
@@ -29,12 +29,28 @@ void force_type(info::device_type &t, const info::device_type &ft) {
...
+      impl = it->second.lock();
+    else {
+      impl = std::make_shared<detail::device_impl>(
+          detail::pi::cast<pi_native_handle>(deviceId),
+          RT::getPlugin<backend::opencl>());
+      if (it == device_impls.end())
+        device_impls[deviceId] = impl;
+      else
+        it->second = impl;
+    }
```
Comment (alexbatashev) [inline sycl/source/device.cpp:52]: Nit
```suggestion
    std::lock_guard<std::mutex> Lock(device_mutex);
    auto It = DeviceImpls.find(DeviceId);
    if (It != DeviceImpls.end() && !It->second.expired())
      impl = It->second.lock();
    else {
      impl = std::make_shared<detail::device_impl>(
          detail::pi::cast<pi_native_handle>(DeviceId),
          RT::getPlugin<backend::opencl>());
      if (it == DeviceImpls.end())
        DeviceImpls[DeviceId] = impl;
      else
        It->second = impl;
    }
```

**Review by alexbatashev — CHANGES_REQUESTED** [review]

File: sycl/include/CL/sycl/device.hpp [hunk 5]
```diff
@@ -192,6 +195,10 @@ class __SYCL_EXPORT device {
   shared_ptr_class<detail::device_impl> impl;
   device(shared_ptr_class<detail::device_impl> impl) : impl(impl) {}
 
+  static std::unordered_map<cl_device_id, std::weak_ptr<detail::device_impl>>
+      device_impls;
```
Comment (romanovvlad) [inline sycl/include/CL/sycl/device.hpp:199]: Is it possible to reuse already existing cache, see platform_impl.hpp:199 (MDeviceCache) ?

Comment (maximdimakov) [reply sycl/include/CL/sycl/device.hpp]: I pushed another solution with reusing an existing code

File: sycl/include/CL/sycl/device.hpp [hunk 6]
```diff
@@ -16,7 +16,6 @@
 #include <CL/sycl/platform.hpp>
 #include <CL/sycl/stl.hpp>
 
-#include <memory>
```
Comment (alexbatashev) [inline sycl/include/CL/sycl/device.hpp:19]: `memory` was actually used for `shared_ptr`.

**Review by alexbatashev — APPROVED** [review]
Otherwise LGTM

File: sycl/include/CL/sycl/device.hpp (hunk 6 above)
Comment (maximdimakov) [reply sycl/include/CL/sycl/device.hpp:19]: It was deleted when I tested map. But memory is included in stl.hpp, that included higher. Do I need to return it?

Comment (bader) [reply sycl/include/CL/sycl/device.hpp:19]: Yes. You can't rely on `CL/sycl/stl.hpp` to include `memory` header for you.

Comment (maximdimakov) [reply sycl/include/CL/sycl/device.hpp]: Returned it

**Review by alexbatashev — APPROVED** [review]

**Review by romanovvlad — APPROVED** [review]

**Pull request merged**